pkg/tls/renewer.go

```diff
@@ -0,0 +1,287 @@
+	rollingUpdateAnnotation string = "update.kyverno.io/force-rolling-update"
+)
+
+// CertRenewer creates rootCA and pem pair to register
+// webhook configurations and webhook server
+// renews RootCA at the given interval
+type CertRenewer struct {
+	client               kubernetes.Interface
+	certRenewalInterval  time.Duration
+	certValidityDuration time.Duration
+	certProps            *certificateProps
+
+	// IP address where Kyverno controller runs. Only required if out-of-cluster.
+	serverIP string
+
+	log logr.Logger
+}
+
+// NewCertRenewer returns an instance of CertRenewer
+func NewCertRenewer(client kubernetes.Interface, clientConfig *rest.Config, certRenewalInterval, certValidityDuration time.Duration, serverIP string, log logr.Logger) (*CertRenewer, error) {
+	certProps, err := newCertificateProps(clientConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &CertRenewer{
+		client:               client,
+		certRenewalInterval:  certRenewalInterval,
+		certValidityDuration: certValidityDuration,
+		certProps:            certProps,
+		serverIP:             serverIP,
+		log:                  log,
+	}, nil
+}
+
+// InitTLSPemPair Loads or creates PEM private key and TLS certificate for webhook server.
+// Created pair is stored in cluster's secret.
+func (c *CertRenewer) InitTLSPemPair() error {
+	logger := c.log.WithName("InitTLSPemPair")
+	ca, err := c.getCASecret()
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	tls, err := c.getTLSSecret()
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	// check they are valid
+	if ca != nil && tls != nil {
+		if validCert(ca, tls, logger) {
+			return nil
+		}
+	}
+	// if not valid, check we can renew them
+	if !IsSecretManagedByKyverno(ca) || !IsSecretManagedByKyverno(tls) {
+		return fmt.Errorf("tls is not valid but certificates are not managed by kyverno, we can't renew them")
+	}
+	// renew them
+	logger.Info("building key/certificate pair for TLS")
+	return c.buildTLSPemPairAndWriteToSecrets(c.serverIP)
+}
+
+// buildTLSPemPairAndWriteToSecrets Issues TLS certificate for webhook server using self-signed CA cert
+// Returns signed and approved TLS certificate in PEM format
+func (c *CertRenewer) buildTLSPemPairAndWriteToSecrets(serverIP string) error {
+	caCert, err := generateCA(c.certValidityDuration)
+	if err != nil {
+		return err
+	}
+	tlsPair, err := generateCert(caCert, c.certProps, serverIP, c.certValidityDuration)
+	if err != nil {
+		return err
+	}
+	if err := c.writeCASecret(caCert); err != nil {
+		return fmt.Errorf("failed to write CA cert to secret: %v", err)
+	}
+	if err = c.writeTLSSecret(tlsPair); err != nil {
+		return fmt.Errorf("unable to save TLS pair to the cluster: %v", err)
+	}
+	return nil
+}
+
+func (c *CertRenewer) getSecret(name string) (*corev1.Secret, error) {
+	if s, err := c.client.CoreV1().Secrets(config.KyvernoNamespace()).Get(context.TODO(), name, metav1.GetOptions{}); err != nil {
```

Comment:
Can we embed a secret lister and use that for `Get`?

Reply:
This one doesn't receive an informer.
Luckily, calls should be pretty rare and using a direct client shouldn't hurt.
I will keep it in mind as an optimization though.

Reply:
OK.